### PR TITLE
#14141 Add externalDocs to @Operation to the JavaSpring generator

### PR DESCRIPTION
**Pull request by arey**
This Pull Request implements the feature describe in https://github.com/OpenAPITools/openapi-generator/issues/14141 with a TDD approach.
It adds an `externalDocs` attribute to the `@Operation` annotation regarding the operation yaml specification.

### PR checklist
 
- [ X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ X] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [ X] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples
- [X ] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.3.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [ X] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.
For JavaSpring comitttee 
@cachescrubber (2022/02) @welshm (2022/02) @MelleD (2022/02) @atextor (2022/02) @manedev79 (2022/02) @javisst (2022/02) @borsch (2022/02) @banlevente (2022/02) @Zomzog (2022/09)
Core Team : @wing328 @jimschubert @cbornet @jmini @etherealjoy @spacether

### CODE REVIEW

**Comment by arey** [comment]
The `ci/circle1:node1` test has crashed with this error:
```
Caused by: java.lang.UnsupportedClassVersionError: org/springframework/boot/maven/RepackageMojo has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 52.0
    at java.lang.ClassLoader.defineClass1 (Native Method)
```

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java
Comment (welshm) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:4481]: You're making changes to `DefaultCodegen` and `CodegenOperation` so you'll also likely need a review from the core team as well (even though this is minor)

File: modules/openapi-generator/src/main/resources/JavaSpring/api.mustache
Comment (welshm) [inline modules/openapi-generator/src/main/resources/JavaSpring/api.mustache:19]: Should this be `#hasExternalDocs` ?

**Review by welshm — COMMENTED** [review]
Can you also regenerate the samples? Or do none of the samples have exeternal docs? If not, perhaps we should add one?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java
Comment (arey) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:4481]: We could maybe initiate the `hasExternalDocs ` attribute at other places. Thus yes the Core Team is welcome :)
@wing328
@jimschubert
@cbornet 
@jmini
@etherealjoy 
@spacether

**Comment by arey** [comment]
> Can you also regenerate the samples? Or do none of the samples have exeternal docs? If not, perhaps we should add one?

I'll we do that. For unit testing I've added an external docs to the `petstore.yaml`. So it should be present in the generated samples. May I have only generate the `spring-boot-oas3.yaml` config?
```
./bin/generate-samples.sh ./bin/configs/spring-boot-oas3.yaml
```

**Comment by welshm** [comment]
> > Can you also regenerate the samples? Or do none of the samples have exeternal docs? If not, perhaps we should add one?
> 
> I'll we do that. For unit testing I've added an external docs to the `petstore.yaml`. So it should be present in the generated samples. May I have only generate the `spring-boot-oas3.yaml` config?
> 
> ```
> ./bin/generate-samples.sh ./bin/configs/spring-boot-oas3.yaml
> ```

You need to run all of the following:

```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
```

Otherwise tests will fail because the diffs won't match.

**Comment by arey** [comment]
I'll first generate the `spring-boot-oas3`. This could me fix some indentation and carriage return issues.
I will now regenerate all the samples.

**Comment by arey** [comment]
Generate all samples command looks like to failed on the master branch:
```
$ ./bin/generate-samples.sh ./bin/generate-samples.sh
...
[main] ERROR o.o.c.config.CodegenConfigurator - Unexpected character ('#' (code 35)): expected a valid value (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: (File); line: 1, column: 2]
[main] ERROR o.o.c.config.CodegenConfigurator - Unexpected character ('#' (code 35)): expected a valid value (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
 at [Source: (File); line: 1, column: 2]
Exception in thread "main" java.lang.RuntimeException: Unable to deserialize config file: ./bin/generate-samples.sh
        at org.openapitools.codegen.config.CodegenConfigurator.readDynamicSettings(CodegenConfigurator.java:167)
        at org.openapitools.codegen.config.CodegenConfigurator.fromFile(CodegenConfigurator.java:89)
        at org.openapitools.codegen.cmd.Generate.execute(Generate.java:293)
        at org.openapitools.codegen.cmd.OpenApiGeneratorCommand.run(OpenApiGeneratorCommand.java:32)
        at org.openapitools.codegen.OpenAPIGenerator.main(OpenAPIGenerator.java:66)
```

**Comment by welshm** [comment]
> Generate all samples command looks like to failed on the master branch:
> 
> ```
> $ ./bin/generate-samples.sh ./bin/generate-samples.sh
> ...
> [main] ERROR o.o.c.config.CodegenConfigurator - Unexpected character ('#' (code 35)): expected a valid value (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
>  at [Source: (File); line: 1, column: 2]
> [main] ERROR o.o.c.config.CodegenConfigurator - Unexpected character ('#' (code 35)): expected a valid value (JSON String, Number, Array, Object or token 'null', 'true' or 'false')
>  at [Source: (File); line: 1, column: 2]
> Exception in thread "main" java.lang.RuntimeException: Unable to deserialize config file: ./bin/generate-samples.sh
>         at org.openapitools.codegen.config.CodegenConfigurator.readDynamicSettings(CodegenConfigurator.java:167)
>         at org.openapitools.codegen.config.CodegenConfigurator.fromFile(CodegenConfigurator.java:89)
>         at org.openapitools.codegen.cmd.Generate.execute(Generate.java:293)
>         at org.openapitools.codegen.cmd.OpenApiGeneratorCommand.run(OpenApiGeneratorCommand.java:32)
>         at org.openapitools.codegen.OpenAPIGenerator.main(OpenAPIGenerator.java:66)
> ```

My guess is there's a serialized config file that needs to be updated because you added `hasExternalDocs` to the config.

**Comment by arey** [comment]
> My guess is there's a serialized config file that needs to be updated because you added hasExternalDocs to the config.

Just a human error.  I've generated then commited all the samples.
The `export_docs_generators.sh` does modified any files.

**Comment by arey** [comment]
Hi @welshm, do you know if the core team members could have a look on my change to  to `DefaultCodegen` and `CodegenOperation` ?

**Comment by welshm** [comment]
> Hi @welshm, do you know if the core team members could have a look on my change to  to `DefaultCodegen` and `CodegenOperation` ?

I'll take another look today and then tag some of them

File: modules/openapi-generator/src/main/resources/JavaSpring/api.mustache
Comment (welshm) [inline modules/openapi-generator/src/main/resources/JavaSpring/api.mustache:176]: Does this also work? I'm not an expert on mustache files, but I think you may not need the `hasExternalDocs` boolean if you check if the parameter `externalDocs` is set like this.

```
}{{/hasAuthMethods}}{{#externalDocs}},
  externalDocs = @ExternalDocumentation(description = "{{externalDocs.description}}", url = "{{externalDocs.url}}")
{{/externalDocs}}
```

Comment (welshm) [reply modules/openapi-generator/src/main/resources/JavaSpring/api.mustache:176]: For example, I see this in the Swift5 templates

```
     {{#externalDocs}}
     - externalDocs: {{.}}
     {{/externalDocs}}
```

**Comment by welshm** [comment]
> Still don't know why the `ci/circleci:node1` is failing: https://circleci.com/gh/OpenAPITools/openapi-generator/56336
> 
> ```
> Caused by: java.lang.UnsupportedClassVersionError: org/springframework/boot/maven/RepackageMojo has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 52.0
> ```
> 
> @welshm does it block my PR?

Shouldn't block your PR

**Review by welshm — APPROVED** [review]
@wing328 will need to review and merge, but this looks great to me!

Thanks for adding the external docs!

**Comment by wing328** [comment]
@arey thanks for the PR but some CI tests failed: https://github.com/OpenAPITools/openapi-generator/actions/runs/3680159465/jobs/6327983250

can you please take a look when you've time?

**Comment by arey** [comment]
Thanks for the review @wing328 
It looks like the `ExternalDocumentation` java import is not generated. I will investigate.

**Comment by arey** [comment]
@wing328 I'll fixed the issue by systematically add the `ExternalDocumentation` java import (like others `io.swagger.v3.oas.annotations.*` imports).
I would have liked to make it conditional. But it requires that `CodegenOperation` shoud be aware of all the `CodegenOperation`. I don't know if other generators do this kind of import optimization?

**Comment by arey** [comment]
Any news?

**Comment by arey** [comment]
Thanks a lot William for this PR. And thank you for all the time you spent on this great project :)

**Pull request merged**
